# Preprocessing

Load World Cup 2018 data from StatsBomb JSON, filter to shots, and create the cleaned shot-level dataset with engineered features (x, y, distance, angle, etc.). Finally, save the result to CSV for use in the modeling notebook.

## Imports and configuration

In [144]:
import pandas as pd
import numpy as np
from pathlib import Path
import json

pd.set_option('display.max_columns', None)


## Load matches and events for World Cup 2018


In [145]:
# Base path to StatsBomb open-data on your machine
BASE = Path(r"C:\\Users\\traik\\Desktop\\Final project data\\open-data-master\\data")

# World Cup 2018: competition_id=43, season_id=3
matches_file = BASE / "matches" / "43" / "3.json"

with open(matches_file, "r", encoding="utf-8") as f:
    matches_wc = json.load(f)

match_ids_wc = [m["match_id"] for m in matches_wc]

events_folder = BASE / "events"
all_events_wc = []

for mid in match_ids_wc:
    fp = events_folder / f"{mid}.json"
    with open(fp, "r", encoding="utf-8") as f:
        all_events_wc.extend(json.load(f))

df_wc = pd.json_normalize(all_events_wc, sep="_")
df_wc.head()


id  index  period     timestamp  minute  \
0  de3be98d-e227-475b-bd55-f57a6a89d308      1       1  00:00:00.000       0   
1  f50ccda4-b768-4f07-9136-8f79fd17dac5      2       1  00:00:00.000       0   
2  b5e98805-0a22-4a5e-a306-7d40651a0f6e      3       1  00:00:00.000       0   
3  762b829f-5f24-4dd7-bfe2-da7e289838bb      4       1  00:00:00.000       0   
4  d4883f20-ce68-4f84-b26a-a049a13cb6be      5       1  00:00:00.240       0   

   second  possession  duration  type_id    type_name  possession_team_id  \
0       0           1     0.000       35  Starting XI                 769   
1       0           1     0.754       35  Starting XI                 769   
2       0           1     9.320       18   Half Start                 769   
3       0           1     9.053       18   Half Start                 769   
4       0           2     0.240       30         Pass                 769   

  possession_team_name  play_pattern_id play_pattern_name  team_id team_name  \
0             Colombia                1      Regular Play      769  Colombia   
1             Colombia                1      Regular Play      768   England   
2             Colombia                1      Regular Play      768   England   
3             Colombia                1      Regular Play      769  Colombia   
4             Colombia                9     From Kick Off      769  Colombia   

   tactics_formation                                     tactics_lineup  \
0              433.0  [{'player': {'id': 4276, 'name': 'David Ospina...   
1              352.0  [{'player': {'id': 3468, 'name': 'Jordan Pickf...   
2                NaN                                                NaN   
3                NaN                                                NaN   
4                NaN                                                NaN   

                           related_events      location  player_id  \
0                                     NaN           NaN        NaN   
1                                     NaN           NaN        NaN   
2  [762b829f-5f24-4dd7-bfe2-da7e289838bb]           NaN        NaN   
3  [b5e98805-0a22-4a5e-a306-7d40651a0f6e]           NaN        NaN   
4  [5fc9acb8-88c3-4cfb-ad9c-fc250c0dffde]  [60.0, 40.0]     3445.0   

                    player_name  position_id        position_name  \
0                           NaN          NaN                  NaN   
1                           NaN          NaN                  NaN   
2                           NaN          NaN                  NaN   
3                           NaN          NaN                  NaN   
4  Radamel Falcao García Zárate         24.0  Left Center Forward   

   pass_recipient_id              pass_recipient_name  pass_length  \
0                NaN                              NaN          NaN   
1                NaN                              NaN          NaN   
2                NaN                              NaN          NaN   
3                NaN                              NaN          NaN   
4             5692.0  Juan Fernando Quintero Paniagua    10.049875   

   pass_angle  pass_height_id pass_height_name pass_end_location  \
0         NaN             NaN              NaN               NaN   
1         NaN             NaN              NaN               NaN   
2         NaN             NaN              NaN               NaN   
3         NaN             NaN              NaN               NaN   
4    3.041924             1.0      Ground Pass      [50.0, 41.0]   

   pass_type_id pass_type_name  pass_body_part_id pass_body_part_name  \
0           NaN            NaN                NaN                 NaN   
1           NaN            NaN                NaN                 NaN   
2           NaN            NaN                NaN                 NaN   
3           NaN            NaN                NaN                 NaN   
4          65.0       Kick Off               38.0           Left Foot   

  under_pressure carry_end_location  pass_outcome_id pas

In [146]:
df_wc['id'].duplicated().sum()


0

## Filter to shots only

In [147]:
df_shots = df_wc[df_wc["type_name"] == "Shot"].copy()
df_shots.shape


(1706, 121)

### Filter out penalty shootouts

In [148]:
df_penalty_shootouts = df_wc[df_wc["period"] == 5]
display(df_penalty_shootouts.shape)
df_penalty_shootouts[["type_name", "shot_type_name", "location", "period"]].head(15)


(94, 121)

type_name shot_type_name       location  period
3995   Half Start            NaN            NaN       5
3996   Half Start            NaN            NaN       5
3997         Shot        Penalty  [109.0, 41.0]       5
3998  Goal Keeper            NaN    [1.0, 40.0]       5
3999         Shot        Penalty  [109.0, 41.0]       5
4000  Goal Keeper            NaN    [1.0, 40.0]       5
4001         Shot        Penalty  [109.0, 41.0]       5
4002  Goal Keeper            NaN    [1.0, 40.0]       5
4003         Shot        Penalty  [109.0, 41.0]       5
4004  Goal Keeper            NaN    [1.0, 40.0]       5
4005         Shot        Penalty  [109.0, 41.0]       5
4006  Goal Keeper            NaN    [1.0, 40.0]       5
4007         Shot        Penalty  [109.0, 41.0]       5
4008  Goal Keeper            NaN    [1.0, 40.0]       5
4009         Shot        Penalty  [109.0, 41.0]       5

In [149]:
shots = df_shots[df_shots["period"] != 5]


## Select relevant columns


In [150]:
cols = [
    "location",
    "counterpress",
    "shot_statsbomb_xg",
    "shot_end_location",
    "shot_type_id",
    "shot_technique_id",
    "shot_outcome_id",
    "shot_body_part_id",
    "shot_open_goal",
    "shot_first_time",
    "shot_one_on_one",
    "shot_aerial_won",
]

shots = shots[cols]
shots.head()


location counterpress  shot_statsbomb_xg   shot_end_location  \
239  [115.0, 18.0]          NaN           0.009816  [120.0, 42.2, 2.0]   
270  [112.0, 54.0]          NaN           0.038204       [113.0, 53.0]   
439   [98.0, 37.0]          NaN           0.045128       [105.0, 37.0]   
548  [119.0, 36.0]          NaN           0.625074  [120.0, 40.5, 3.1]   
815   [97.0, 56.0]          NaN           0.021760       [100.0, 54.0]   

     shot_type_id  shot_technique_id  shot_outcome_id  shot_body_part_id  \
239          62.0               93.0            100.0               40.0   
270          87.0               93.0             96.0               40.0   
439          87.0               93.0             96.0               40.0   
548          87.0               93.0             98.0               37.0   
815          87.0               93.0             96.0               40.0   

    shot_open_goal shot_first_time shot_one_on_one shot_aerial_won  
239            NaN             NaN             NaN             NaN  
270            NaN             NaN             NaN             NaN  
439            NaN             NaN             NaN             NaN  
548           True             NaN             NaN             NaN  
815            NaN             NaN             NaN             NaN

## Drop unreliable columns



In [151]:
shots.isna().sum()

location                0
counterpress         1667
shot_statsbomb_xg       0
shot_end_location       0
shot_type_id            0
shot_technique_id       0
shot_outcome_id         0
shot_body_part_id       0
shot_open_goal       1650
shot_first_time      1310
shot_one_on_one      1612
shot_aerial_won      1543
dtype: int64

In [152]:
# Drop counterpress

shots.drop(["counterpress"], axis=1, inplace=True)

shots.isna().sum()


location                0
shot_statsbomb_xg       0
shot_end_location       0
shot_type_id            0
shot_technique_id       0
shot_outcome_id         0
shot_body_part_id       0
shot_open_goal       1650
shot_first_time      1310
shot_one_on_one      1612
shot_aerial_won      1543
dtype: int64

In [153]:
# Drop shot_end_location as execution of shot should not play a role in measuring the quality of shot.

shots.drop(["shot_end_location"], axis=1, inplace=True)

## Clean boolean shot flags

`shot_open_goal`, `shot_first_time`, `shot_one_on_one`, `shot_aerial_won` are mapped from `NaN` → 0 and cast to integer 0/1.

In [154]:
binary_cols = [
    "shot_open_goal",
    "shot_first_time",
    "shot_one_on_one",
    "shot_aerial_won",
]

for col in binary_cols:
    shots[col] = shots[col].fillna(0).astype(int)

shots[binary_cols].head()


shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won
239               0                0                0                0
270               0                0                0                0
439               0                0                0                0
548               1                0                0                0
815               0                0                0                0

## Extract shot location (x, y)

We convert the StatsBomb `location` list `[x, y]` into separate numeric `x` and `y` columns, then drop the original `location` column.

In [155]:
shots["x"] = shots["location"].apply(lambda loc: loc[0])
shots["y"] = shots["location"].apply(lambda loc: loc[1])

shots.drop("location", axis=1, inplace=True)
shots.head()


shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_outcome_id  \
239           0.009816          62.0               93.0            100.0   
270           0.038204          87.0               93.0             96.0   
439           0.045128          87.0               93.0             96.0   
548           0.625074          87.0               93.0             98.0   
815           0.021760          87.0               93.0             96.0   

     shot_body_part_id  shot_open_goal  shot_first_time  shot_one_on_one  \
239               40.0               0                0                0   
270               40.0               0                0                0   
439               40.0               0                0                0   
548               37.0               1                0                0   
815               40.0               0                0                0   

     shot_aerial_won      x     y  
239                0  115.0  18.0  
270                0  112.0  54.0  
439                0   98.0  37.0  
548                0  119.0  36.0  
815                0   97.0  56.0

## Create goal label `is_goal`

We use your mapping: `shot_outcome_id == 97` corresponds to goals in this dataset.

In [156]:
shots["is_goal"] = (shots["shot_outcome_id"] == 97).astype(int)
shots["is_goal"].value_counts()


is_goal
0    1510
1     157
Name: count, dtype: int64

In [157]:
# Drop the outcome_id now that we have the label
shots.drop("shot_outcome_id", axis=1, inplace=True)
shots.head()


shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
239           0.009816          62.0               93.0               40.0   
270           0.038204          87.0               93.0               40.0   
439           0.045128          87.0               93.0               40.0   
548           0.625074          87.0               93.0               37.0   
815           0.021760          87.0               93.0               40.0   

     shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won      x  \
239               0                0                0                0  115.0   
270               0                0                0                0  112.0   
439               0                0                0                0   98.0   
548               1                0                0                0  119.0   
815               0                0                0                0   97.0   

        y  is_goal  
239  18.0        0  
270  54.0        0  
439  37.0        0  
548  36.0        0  
815  56.0        0

## Add geometry features: distance and angle

We compute:
- `distance`: distance from shot location to the centre of the goal
- `angle`: angle between lines from the shot location to the two goalposts

Pitch and goal coordinates follow the StatsBomb convention: x ∈ [0,120], y ∈ [0,80], goal centered at (120, 40).

In [158]:
# Goal coordinates (StatsBomb pitch)
goal_x = 120
goal_y = 40
left_post_y = 36.8
right_post_y = 43.2

# Distance to goal centre
shots["distance"] = np.sqrt((goal_x - shots["x"])**2 + (goal_y - shots["y"])**2)

def calc_angle(row):
    x = row["x"]
    y = row["y"]
    angle_left = np.arctan2(left_post_y - y, goal_x - x)
    angle_right = np.arctan2(right_post_y - y, goal_x - x)
    return abs(angle_right - angle_left)



## Create final featured shots dataframe

We drop `shot_end_location` so the features only describe the chance **before** the shot outcome. This final dataframe `shots_featured` will be saved to CSV and used in the modeling notebook.

In [159]:
shots_featured = shots.copy()

shots_featured["angle"] = shots.apply(calc_angle, axis=1)
shots_featured[["x", "y", "distance", "angle"]].head()


shots_featured.head()


shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
239           0.009816          62.0               93.0               40.0   
270           0.038204          87.0               93.0               40.0   
439           0.045128          87.0               93.0               40.0   
548           0.625074          87.0               93.0               37.0   
815           0.021760          87.0               93.0               40.0   

     shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won      x  \
239               0                0                0                0  115.0   
270               0                0                0                0  112.0   
439               0                0                0                0   98.0   
548               1                0                0                0  119.0   
815               0                0                0                0   97.0   

        y  is_goal   distance     angle  
239  18.0        0  22.561028  0.064071  
270  54.0        0  16.124515  0.202196  
439  37.0        0  22.203603  0.283785  
548  36.0        0   4.123106  0.758049  
815  56.0        0  28.017851  0.187756

### Checking for duplicate columns

In [160]:
shots_featured.duplicated().sum()

22

In [161]:
dupes = shots_featured[shots_featured.duplicated(keep=False)]
dupes.head(60)



shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
1972               0.7835          88.0               93.0               40.0   
13367              0.7835          88.0               93.0               40.0   
14420              0.7835          88.0               93.0               40.0   
22589              0.7835          88.0               93.0               40.0   
23240              0.7835          88.0               93.0               40.0   
60457              0.7835          88.0               93.0               40.0   
61587              0.7835          88.0               93.0               40.0   
74316              0.7835          88.0               93.0               40.0   
77105              0.7835          88.0               93.0               38.0   
90799              0.7835          88.0               93.0               40.0   
93819              0.7835          88.0               93.0               40.0   
111493             0.7835          88.0               93.0               38.0   
113729             0.7835          88.0               93.0               38.0   
113793             0.7835          88.0               93.0               40.0   
123149             0.7835          88.0               93.0               40.0   
140041             0.7835          88.0               93.0               38.0   
144899             0.7835          88.0               93.0               40.0   
147725             0.7835          88.0               93.0               38.0   
150644             0.7835          88.0               93.0               40.0   
154140             0.7835          88.0               93.0               40.0   
158923             0.7835          88.0               93.0               40.0   
170074             0.7835          88.0               93.0               40.0   
190410             0.7835          88.0               93.0               40.0   
198494             0.7835          88.0               93.0               40.0   
212467             0.7835          88.0               93.0               38.0   
225714             0.7835          88.0               93.0               40.0   
225870             0.7835          88.0               93.0               38.0   

        shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won  \
1972                 0                0                0                0   
13367                0                0                0                0   
14420                0                0                0                0   
22589                0                0                0                0   
23240                0                0                0                0   
60457                0                0                0                0   
61587                0                0                0                0   
74316                0                0                0                0   
77105                0                0                0                0   
90799                0                0                0                0   
93819                0                0                0                0   
111493               0                0                0                0   
113729               0                0                0                0   
113793               0                0                0                0   
123149               0                0                0                0   
140041               0                0                0                0   
144899               0                0                0                0   
147725               0                0                0                0   
150644               0                0                0                0   
154140               0                0                0                0   
158923               0                0                0                0   
170074               0                0      

In [162]:
# duplicates are mostly penalty shots. we dont drop them.


In [163]:
shots_featured.iloc[32:45,:]

shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
4454           0.257208          87.0               93.0               40.0   
4540           0.006950          87.0               93.0               40.0   
4604           0.173503          87.0               93.0               37.0   
5209           0.037858          87.0               93.0               40.0   
5215           0.068467          87.0               93.0               40.0   
5228           0.019955          87.0               93.0               40.0   
5497           0.085507          87.0               93.0               37.0   
5511           0.024078          87.0               93.0               38.0   
5516           0.101480          87.0               90.0               37.0   
5559           0.071149          87.0               93.0               37.0   
5616           0.054441          87.0               93.0               40.0   
5693           0.048023          87.0               95.0               40.0   
6169           0.070812          87.0               93.0               40.0   

      shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won  \
4454               0                0                0                0   
4540               0                0                0                0   
4604               0                0                0                0   
5209               0                0                0                0   
5215               0                0                0                0   
5228               0                0                0                0   
5497               0                0                0                0   
5511               0                0                0                0   
5516               0                0                0                0   
5559               0                0                0                1   
5616               0                1                0                0   
5693               0                1                0                0   
6169               0                0                0                0   

          x     y  is_goal   distance     angle  
4454  117.0  40.0        0   3.000000  1.635290  
4540   91.0  20.0        0  35.227830  0.149673  
4604  112.0  42.0        0   8.246211  0.725265  
5209  109.0  55.0        0  18.601075  0.206680  
5215  107.0  33.0        0  14.764823  0.380905  
5228   91.0  49.0        0  30.364453  0.200819  
5497  113.0  40.0        0   7.000000  0.857556  
5511   90.0  41.0        0  30.016662  0.212298  
5516  111.0  43.0        0   9.486833  0.625448  
5559  109.0  35.0        0  12.083046  0.478387  
5616   99.0  34.0        0  21.840330  0.280358  
5693  108.0  27.0        0  17.691806  0.248426  
6169  107.0  27.0        0  18.384776  0.248594

In [164]:
shots_featured.shape

(1667, 13)

In [165]:
shots_featured.is_goal.value_counts()

is_goal
0    1510
1     157
Name: count, dtype: int64

## Save preprocessed shots to CSV

In [166]:
output_path = Path("shots_featured_wc2018.csv")
shots_featured.to_csv(output_path, index=False)
print(f"Saved preprocessed shots to {output_path.resolve()}")


Saved preprocessed shots to C:\Users\traik\Desktop\ML_Bootcamp\Final_Project\Final_Project_Ironhack\shots_featured_wc2018.csv


# Loading Euro 2020 Data to test models on

In [167]:


# EURO 2020
matches_file_euro = BASE / "matches" / "55" / "43.json"
with open(matches_file_euro, "r", encoding="utf-8") as f:
    matches_euro = json.load(f)

match_ids_euro = [m["match_id"] for m in matches_euro]

# Load events
events_folder = BASE / "events"
all_events_euro = []

for mid in match_ids_euro:
    fp = events_folder / f"{mid}.json"
    with open(fp, "r", encoding="utf-8") as f:
        all_events_euro.extend(json.load(f))

df_euro = pd.json_normalize(all_events_euro, sep="_")
df_euro.head()

id  index  period     timestamp  minute  \
0  2e4b1b7f-e67e-4bef-8cf1-2eab2ab81af2      1       1  00:00:00.000       0   
1  f73c3f38-8c45-4802-8cb9-5a87234f385b      2       1  00:00:00.000       0   
2  9098375d-3f07-4241-bdc9-3602a2e5c754      3       1  00:00:00.000       0   
3  18099197-a92a-4398-ac5e-39dad539c3eb      4       1  00:00:00.000       0   
4  ea57ee78-a4ef-4d62-81c7-98cdf1c11925      5       1  00:00:00.967       0   

   second  possession  duration  type_id    type_name  possession_team_id  \
0       0           1  0.000000       35  Starting XI                 773   
1       0           1  0.000000       35  Starting XI                 773   
2       0           1  0.000000       18   Half Start                 773   
3       0           1  0.000000       18   Half Start                 773   
4       0           2  1.511008       30         Pass                 772   

  possession_team_name  play_pattern_id play_pattern_name  team_id  \
0          Switzerland                1      Regular Play      773   
1          Switzerland                1      Regular Play      772   
2          Switzerland                1      Regular Play      772   
3          Switzerland                1      Regular Play      773   
4                Spain                9     From Kick Off      772   

     team_name  tactics_formation  \
0  Switzerland             4231.0   
1        Spain              433.0   
2        Spain                NaN   
3  Switzerland                NaN   
4        Spain                NaN   

                                      tactics_lineup  \
0  [{'player': {'id': 5550, 'name': 'Yann Sommer'...   
1  [{'player': {'id': 11748, 'name': 'Unai Simón ...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                           related_events      location  player_id  \
0                                     NaN           NaN        NaN   
1                                     NaN           NaN        NaN   
2  [18099197-a92a-4398-ac5e-39dad539c3eb]           NaN        NaN   
3  [9098375d-3f07-4241-bdc9-3602a2e5c754]           NaN        NaN   
4  [66655fd3-2764-4532-b1bb-0a01074fed30]  [61.0, 40.1]     3477.0   

                  player_name  position_id   position_name  pass_recipient_id  \
0                         NaN          NaN             NaN                NaN   
1                         NaN          NaN             NaN                NaN   
2                         NaN          NaN             NaN                NaN   
3                         NaN          NaN             NaN                NaN   
4  Álvaro Borja Morata Martín         23.0  Center Forward             6892.0   

    pass_recipient_name  pass_length  pass_angle  pass_height_id  \
0                   NaN          NaN         NaN             NaN   
1                   NaN          NaN         NaN             NaN   
2                   NaN          NaN         NaN             NaN   
3                   NaN          NaN         NaN             NaN   
4  Pau Francisco Torres     26.68108   -2.883887             1.0   

  pass_height_name pass_end_location  pass_body_part_id pass_body_part_name  \
0              NaN               NaN                NaN                 NaN   
1              NaN               NaN                NaN                 NaN   
2              NaN               NaN                NaN                 NaN   
3              NaN               NaN                NaN                 NaN   
4      Ground Pass      [35.2, 33.3]               40.0          Right Foot   

   pass_type_id pass_type_name carry_end_location under_pressure  \
0           NaN            NaN                NaN            NaN   
1           NaN            NaN                NaN            NaN   
2           NaN            NaN                NaN            NaN   
3           NaN            NaN                Na

### checking for duplicates in original data frame

In [168]:
df_euro["id"].duplicated().sum()

0

# Filter Data Frame to type shot

In [169]:
df_shots_euro = df_euro[df_euro["type_name"] == "Shot"].copy()
df_shots_euro.shape

(1289, 142)

remove penalty shootouts

In [170]:
df_penalty_shootouts_euro = df_euro[df_euro["period"] == 5]
display(df_penalty_shootouts_euro.shape)
df_penalty_shootouts_euro[["type_name", "shot_type_name", "location", "period"]].head(15)


(92, 142)

type_name shot_type_name       location  period
4968   Half Start            NaN            NaN       5
4969   Half Start            NaN            NaN       5
4970         Shot        Penalty  [108.0, 40.0]       5
4971  Goal Keeper            NaN    [1.0, 40.0]       5
4972         Shot        Penalty  [108.0, 40.0]       5
4973  Goal Keeper            NaN    [1.0, 40.0]       5
4974         Shot        Penalty  [108.0, 40.0]       5
4975  Goal Keeper            NaN    [1.0, 40.0]       5
4976         Shot        Penalty  [108.0, 40.0]       5
4977  Goal Keeper            NaN    [1.0, 40.0]       5
4978         Shot        Penalty  [108.0, 40.0]       5
4979  Goal Keeper            NaN    [1.0, 40.0]       5
4980         Shot        Penalty  [108.0, 40.0]       5
4981  Goal Keeper            NaN    [1.0, 40.0]       5
4982         Shot        Penalty  [108.0, 40.0]       5

In [171]:
shots_euro = df_shots_euro[df_shots_euro["period"] != 5]
shots_euro.shape


(1251, 142)

Euro data set will only be used for evaluation. Thus, the same feature and pre-processing steps will be maintained and applied as before.

In [172]:
cols = [
    "location",
    "counterpress",
    "shot_statsbomb_xg",
    "shot_end_location",
    "shot_type_id",
    "shot_technique_id",
    "shot_outcome_id",
    "shot_body_part_id",
    "shot_open_goal",
    "shot_first_time",
    "shot_one_on_one",
    "shot_aerial_won",
]

shots_euro = shots_euro[cols]
display(shots_euro.head())
shots_euro.shape


location counterpress  shot_statsbomb_xg   shot_end_location  \
72     [91.9, 39.0]          NaN           0.030169        [99.7, 40.0]   
400    [96.5, 29.7]          NaN           0.021315       [110.2, 37.8]   
748    [97.8, 25.9]          NaN           0.047888  [120.0, 38.3, 3.5]   
1016  [112.0, 40.7]          NaN           0.100356  [118.8, 38.7, 2.2]   
1449   [99.7, 58.6]          NaN           0.008758       [100.7, 57.7]   

      shot_type_id  shot_technique_id  shot_outcome_id  shot_body_part_id  \
72            87.0               93.0             96.0               38.0   
400           87.0               91.0             96.0               38.0   
748           62.0               93.0             98.0               40.0   
1016          87.0               93.0            100.0               37.0   
1449          87.0               93.0             96.0               38.0   

     shot_open_goal shot_first_time shot_one_on_one shot_aerial_won  
72              NaN             NaN             NaN             NaN  
400             NaN            True             NaN             NaN  
748             NaN             NaN             NaN             NaN  
1016            NaN             NaN             NaN             NaN  
1449            NaN             NaN             NaN             NaN

(1251, 12)

In [173]:
display(shots_euro.isna().sum())
shots_euro.shape

location                0
counterpress         1251
shot_statsbomb_xg       0
shot_end_location       0
shot_type_id            0
shot_technique_id       0
shot_outcome_id         0
shot_body_part_id       0
shot_open_goal       1236
shot_first_time       857
shot_one_on_one      1206
shot_aerial_won      1117
dtype: int64

(1251, 12)

In [174]:
# Drop counterpress

shots_euro.drop(["counterpress"], axis=1, inplace=True)

shots_euro.isna().sum()


location                0
shot_statsbomb_xg       0
shot_end_location       0
shot_type_id            0
shot_technique_id       0
shot_outcome_id         0
shot_body_part_id       0
shot_open_goal       1236
shot_first_time       857
shot_one_on_one      1206
shot_aerial_won      1117
dtype: int64

In [175]:
# Drop Shot_end_location
shots_euro.drop(["shot_end_location"], axis=1, inplace=True)


In [176]:
display(shots_euro.shape)
shots.shape


(1251, 10)

(1667, 12)

In [177]:
# map false to 0 for binary columns. apply astype(int)

binary_cols = [
    "shot_open_goal",
    "shot_first_time",
    "shot_one_on_one",
    "shot_aerial_won",
]

for col in binary_cols:
    shots_euro[col] = shots_euro[col].fillna(0).astype(int)

display(shots_euro[binary_cols].head())

shots_euro.isna().sum()



shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won
72                 0                0                0                0
400                0                1                0                0
748                0                0                0                0
1016               0                0                0                0
1449               0                0                0                0

location             0
shot_statsbomb_xg    0
shot_type_id         0
shot_technique_id    0
shot_outcome_id      0
shot_body_part_id    0
shot_open_goal       0
shot_first_time      0
shot_one_on_one      0
shot_aerial_won      0
dtype: int64

In [178]:
# Extracting shot locations

shots_euro["x"] = shots_euro["location"].apply(lambda loc: loc[0])
shots_euro["y"] = shots_euro["location"].apply(lambda loc: loc[1])

shots_euro.drop("location", axis=1, inplace=True)
display(shots_euro.head())
shots_euro.shape


shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_outcome_id  \
72             0.030169          87.0               93.0             96.0   
400            0.021315          87.0               91.0             96.0   
748            0.047888          62.0               93.0             98.0   
1016           0.100356          87.0               93.0            100.0   
1449           0.008758          87.0               93.0             96.0   

      shot_body_part_id  shot_open_goal  shot_first_time  shot_one_on_one  \
72                 38.0               0                0                0   
400                38.0               0                1                0   
748                40.0               0                0                0   
1016               37.0               0                0                0   
1449               38.0               0                0                0   

      shot_aerial_won      x     y  
72                  0   91.9  39.0  
400                 0   96.5  29.7  
748                 0   97.8  25.9  
1016                0  112.0  40.7  
1449                0   99.7  58.6

(1251, 11)

In [179]:
# Replacing shou_outcome with is_goal feature

shots_euro["is_goal"] = (shots_euro["shot_outcome_id"] == 97).astype(int)
shots_euro["is_goal"].value_counts()


is_goal
0    1120
1     131
Name: count, dtype: int64

In [180]:
# Drop the outcome_id now that we have the label
shots_euro.drop("shot_outcome_id", axis=1, inplace=True)
display(shots_euro.head())
shots_euro.shape


shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
72             0.030169          87.0               93.0               38.0   
400            0.021315          87.0               91.0               38.0   
748            0.047888          62.0               93.0               40.0   
1016           0.100356          87.0               93.0               37.0   
1449           0.008758          87.0               93.0               38.0   

      shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won  \
72                 0                0                0                0   
400                0                1                0                0   
748                0                0                0                0   
1016               0                0                0                0   
1449               0                0                0                0   

          x     y  is_goal  
72     91.9  39.0        0  
400    96.5  29.7        0  
748    97.8  25.9        0  
1016  112.0  40.7        0  
1449   99.7  58.6        0

(1251, 11)

In [181]:
# Adding geometry features

# Distance to goal centre

shots_euro["distance"] = np.sqrt((goal_x - shots_euro["x"])**2 + (goal_y - shots_euro["y"])**2)

shots_euro["angle"] = shots_euro.apply(calc_angle,axis=1)

In [182]:
shots_euro.head()

shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
72             0.030169          87.0               93.0               38.0   
400            0.021315          87.0               91.0               38.0   
748            0.047888          62.0               93.0               40.0   
1016           0.100356          87.0               93.0               37.0   
1449           0.008758          87.0               93.0               38.0   

      shot_open_goal  shot_first_time  shot_one_on_one  shot_aerial_won  \
72                 0                0                0                0   
400                0                1                0                0   
748                0                0                0                0   
1016               0                0                0                0   
1449               0                0                0                0   

          x     y  is_goal   distance     angle  
72     91.9  39.0        0  28.117788  0.226500  
400    96.5  29.7        0  25.658137  0.228027  
748    97.8  25.9        0  26.299240  0.205564  
1016  112.0  40.7        0   8.030567  0.756483  
1449   99.7  58.6        0  27.532708  0.172017

In [183]:
display(shots_featured.shape)
display(shots.shape)
shots_euro.shape

(1667, 13)

(1667, 12)

(1251, 13)

In [184]:
display(shots_featured.columns)
display(shots.columns)
display(shots_euro.columns)

Index(['shot_statsbomb_xg', 'shot_type_id', 'shot_technique_id',
       'shot_body_part_id', 'shot_open_goal', 'shot_first_time',
       'shot_one_on_one', 'shot_aerial_won', 'x', 'y', 'is_goal', 'distance',
       'angle'],
      dtype='object')

Index(['shot_statsbomb_xg', 'shot_type_id', 'shot_technique_id',
       'shot_body_part_id', 'shot_open_goal', 'shot_first_time',
       'shot_one_on_one', 'shot_aerial_won', 'x', 'y', 'is_goal', 'distance'],
      dtype='object')

Index(['shot_statsbomb_xg', 'shot_type_id', 'shot_technique_id',
       'shot_body_part_id', 'shot_open_goal', 'shot_first_time',
       'shot_one_on_one', 'shot_aerial_won', 'x', 'y', 'is_goal', 'distance',
       'angle'],
      dtype='object')

In [185]:
# checking for duplicates before saving file

shots_euro.duplicated().sum()

6

In [186]:
shots_euro.duplicated

<bound method DataFrame.duplicated of         shot_statsbomb_xg  shot_type_id  shot_technique_id  shot_body_part_id  \
72               0.030169          87.0               93.0               38.0   
400              0.021315          87.0               91.0               38.0   
748              0.047888          62.0               93.0               40.0   
1016             0.100356          87.0               93.0               37.0   
1449             0.008758          87.0               93.0               38.0   
...                   ...           ...                ...                ...   
192306           0.030229          87.0               93.0               38.0   
192352           0.043607          87.0               93.0               40.0   
192386           0.025420          87.0               91.0               40.0   
192531           0.058556          87.0               91.0               40.0   
192561           0.127872          87.0               93.0             

These are not duplicates but penalty shots taken in games. Penatly shootouts are already removed

# Save Euro 2020 Data to csv file


In [187]:
# save data frame to csv file

output_path = Path("shots_featured_eu2020.csv")
shots_euro.to_csv(output_path, index=False)
print(f"Saved Euro preprocessed shots to {output_path.resolve()}")


Saved Euro preprocessed shots to C:\Users\traik\Desktop\ML_Bootcamp\Final_Project\Final_Project_Ironhack\shots_featured_eu2020.csv
